tools/coq_dune.ml

```diff
@@ -0,0 +1,210 @@
+  let eflag, edep = if List.tl dir = ["Init"] then "-noinit -R theories Coq ", [] else "", ["theories/Init/Prelude.vo"] in
+  (* Correct path from global to local "theories/Init/Decimal.vo" -> "../../theories/Init/Decimal.vo" *)
+  let deps = List.map (fun s -> sdir ^ s) (edep @ vo.deps) in
+  (* The source file is also corrected as we will call coqtop from the top dir *)
+  let source = String.concat "/" dir ^ "/" ^ Filename.(remove_extension vo.target) ^ ".v" in
+  (* The final build rule *)
+  let action = sprintf "(chdir %%{project_root} (run coqtop -boot %s-compile %s))" eflag source in
+  pp_rule fmt [vo.target] deps action
+
+let pp_ml4_dep _dir fmt ml =
+  let target = Filename.(remove_extension ml) ^ ".ml" in
+  let ml4_rule = "(run coqp5 -loc loc -impl %{pp-file} -o %{targets})" in
+  pp_rule fmt [target] [ml] ml4_rule
+
+let pp_mlg_dep _dir fmt ml =
+  let target = Filename.(remove_extension ml) ^ ".ml" in
+  let ml4_rule = "(run coqpp %{pp-file})" in
+  pp_rule fmt [target] [ml] ml4_rule
+
+let pp_dep dir fmt oo = match oo with
+  | VO vo -> pp_vo_dep dir fmt vo
+  | ML4 f -> pp_ml4_dep dir fmt f
+  | MLG f -> pp_mlg_dep dir fmt f
+
+let out_install fmt dir ff =
+  let itarget = String.concat "/" dir in
+  let ff = pmap (function | VO vo -> Some vo.target | _ -> None) ff in
+  let pp_ispec fmt tg = fprintf fmt "(%s as %s)" tg (itarget^"/"^tg) in
+  fprintf fmt "(install@\n @[(section lib)@\n(files @[%a@])@])@\n"
+    (pp_list pp_ispec sep) ff
+
+(* For each directory, we must record two things, the build rules and
+   the install specification. *)
+let record_dune (d,ff) =
+  let sd = String.concat "/" d in
+  if Sys.file_exists sd && Sys.is_directory sd then
+    let out = open_out (sd^"/dune") in
+    let fmt = formatter_of_out_channel out in
+    if List.nth d 0 = "plugins" then
+      fprintf fmt "(include plugin_base.dune)@\n";
+    out_install fmt d ff;
+    List.iter (pp_dep d fmt) ff;
+    fprintf fmt "%!";
+    close_out out
+  else
+    eprintf "error in %s@\n%!" sd
+
+(*  *)
+let choose_ml4g_form f =
+  if Filename.check_suffix f ".ml4" then ML4 f
+  else MLG f
+
+let scan_mlg4 m d =
+  let dir = ["plugins"; d] in
+  let m = DirMap.add dir [] m in
+  let ml4 = Sys.(List.filter (fun f -> Filename.(check_suffix f ".ml4" || check_suffix f ".mlg"))
+                   Array.(to_list @@ readdir String.(concat "/" dir))) in
+  List.fold_left (fun m f -> add_map_list ["plugins"; d] (choose_ml4g_form f) m) m ml4
+
+let scan_plugins m =
+  let dirs = Sys.(List.filter (fun f -> is_directory @@ "plugins/"^f) Array.(to_list @@ readdir "plugins/")) in
+  List.fold_left scan_mlg4 m dirs
+
+(* Process .vfiles.d and generate a skeleton for the dune file *)
+let parse_coqdep_line l =
+  match Str.(split (regexp ":") l) with
+  | [targets;deps] ->
+    let targets = Str.(split (regexp "[ \t]+") targets) in
+    let deps    = Str.(split (regexp "[ \t]+") deps) in
+    let targets = filter_no_vo targets in
```

Comment:
fix the indentation

Reply:
I will, but honestly I find this much easier to read.